projects/concurrency/main.go

```diff
@@ -0,0 +1,111 @@
+package main
+
+import "fmt"
+
+type Value[K comparable] struct {
+	data any
+	next *K
+	prev *K
+}
+
+type Cache[K comparable] struct {
+	size  int
+	m     map[K]Value[K]
+	first *K
+	last  *K
+}
+
+func NewCache[K comparable](entryLimit int) Cache[K] { //All K should be unique && same type && comparable type like primitive that can compare with each other
+	return Cache[K]{
+		size:  entryLimit,
+		m:     make(map[K]Value[K]),
+		first: nil,
+		last:  nil,
+	}
+}
+
+func (c *Cache[K]) Put(key K, value Value[K]) bool {
+
+	if currentValue, isExisted := c.m[key]; isExisted {
+		nextKey := *currentValue.next
+		nextValue := c.m[nextKey]
+
+		if *c.first == key {
+			nextValue.prev = nil
+			*c.first = nextKey
+		} else {
+			nextValue.prev = currentValue.prev
+			prevKey := *currentValue.prev
+			prevValue := c.m[prevKey]
+			prevValue.next = currentValue.next
+			c.m[prevKey] = prevValue
+		}
+		c.m[nextKey] = nextValue
+		lastValue := c.m[*c.last]
+		lastValue.next = &key
+		c.m[*c.last] = lastValue
+		currentValue.prev = c.last
+		currentValue.next = nil
+		currentValue.data = value.data
+		c.m[key] = currentValue
+		c.last = &key
+		return true
+	} else if len(c.m) < c.size {
+
+		if c.first == nil {
+			value.prev = nil
+			c.first = &key
+		} else {
+			oldLastValue := c.m[*c.last]
+			oldLastValue.next = &key
+			c.m[*c.last] = oldLastValue
+			value.prev = c.last
+		}
+		value.next = nil
+		c.last = &key
+		c.m[key] = value
+		return false
+	} else {
+		currentLastValue := c.m[*c.last]
+		currentLastValue.next = &key
+		c.m[*c.last] = currentLastValue
+
+		currentFirstValue := c.m[*c.first]
+		delete(c.m, *c.first)
+		c.first = currentFirstValue.next
+
+		value.prev = c.last
+		value.next = nil
+		c.m[key] = value
+
+		c.last = &key
+		return false
+	}
+}
+
+func (c *Cache[K]) Get(key K) (*Value[K], bool) {
```

Comment:
Get should also make the key be the most recently used. i.e. if you add 10 values, then get one, the other un-read 9 values should be evicted (if necessary) before the one gotten one.